src/util/rocksdb/properties.rs

```diff
@@ -466,6 +467,188 @@ impl DecodeProperties for UserCollectedProperties {
+        let mut props = UserProperties::new();
+        props.encode(PROP_RANGE_INDEX, buf);
+        props
+    }
+
+    pub fn decode<T: DecodeProperties>(props: &T) -> Result<RangeProperties> {
+        match RangeProperties::decode_from_range_properties(props) {
+            Ok(res)=>return Ok(res),
+            Err(e)=>debug!("old_version:can not decode to RangeProperties:{:?}, try to decode from SizeProperties",e),
+        }
+        SizeProperties::decode(props).map(|res| res.into())
+    }
+
+    fn decode_from_range_properties<T: DecodeProperties>(props: &T) -> Result<RangeProperties> {
+        let mut res = RangeProperties::default();
+        let mut buf = props.decode(PROP_RANGE_INDEX)?;
+        while !buf.is_empty() {
+            let klen = number::decode_u64(&mut buf)?;
+            let mut k = vec![0; klen as usize];
+            buf.read_exact(&mut k)?;
+            let mut offsets = OffsetsInRange::default();
+            offsets.size = number::decode_u64(&mut buf)?;
+            offsets.keys = number::decode_u64(&mut buf)?;
+            res.offsets.insert(k, offsets);
+        }
+        Ok(res)
+    }
+
+    pub fn get_approximate_size_in_range(&self, start: &[u8], end: &[u8]) -> u64 {
+        self.get_approximate_distance_in_range(OffsetKindInRange::Size, start, end)
+    }
+
+    pub fn get_approximate_keys_in_range(&self, start: &[u8], end: &[u8]) -> u64 {
+        self.get_approximate_distance_in_range(OffsetKindInRange::Keys, start, end)
+    }
+
+    fn get_approximate_distance_in_range(
+        &self,
+        kind: OffsetKindInRange,
+        start: &[u8],
+        end: &[u8],
+    ) -> u64 {
+        assert!(start <= end);
+        if start == end {
+            return 0;
+        }
+
+        let range = self.offsets.range::<[u8], _>((Unbounded, Included(start)));
+        let start_offset = match range.last() {
+            Some((_, v)) => v.get(kind),
+            None => 0,
+        };
+        let mut range = self.offsets.range::<[u8], _>((Included(end), Unbounded));
+        let end_offset = match range.next() {
+            Some((_, v)) => v.get(kind),
+            None => self.offsets.iter().last().map_or(0, |(_, v)| v.get(kind)),
+        };
+        if end_offset < start_offset {
+            panic!(
+                "start {:?} end {:?} start_offset {} end_offset {}",
+                start, end, start_offset, end_offset
+            );
+        }
+        end_offset - start_offset
+    }
+
+    pub fn smallest_key(&self) -> Option<Vec<u8>> {
+        self.offsets.iter().next().map(|(key, _)| key.clone())
+    }
+
+    pub fn largest_key(&self) -> Option<Vec<u8>> {
+        self.offsets.iter().last().map(|(key, _)| key.clone())
+    }
+}
+
+impl From<SizeProperties> for RangeProperties {
+    fn from(p: SizeProperties) -> RangeProperties {
+        let mut res = RangeProperties::default();
+        for (key, size_handle) in p.index_handles.0 {
+            let mut range = OffsetsInRange::default();
```

Comment:
When we upgrade from an existing cluster, the "keys" field will be zero or a smaller value, will that be a problem?
For example, we may report a wrong approximate number of keys to PD.

Reply:
no, a smaller `keys` will not cause any change, only a bigger one may cause the split operation.